extensions/wikia/AdEngine/js/context/adContext.js

```diff
@@ -42,11 +42,67 @@ define('ext.wikia.adEngine.adContext', [
 		return context.targeting.pageType === pageType;
 	}
 
+	function isPageFairDetectionEnabled() {
+		var isSupportedGeo = geo.isProperGeo(instantGlobals.wgAdDriverPageFairDetectionCountries);
+		return isUrlParamSet('pagefairdetection') || (isSupportedGeo && sampler.sample('pageFairDetection', 1, 10));
+	}
+
+	function isGSCEnabled() {
+		return abTest.getGroup('PROJECT_43') === 'GROUP_5' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverGoogleConsumerSurveysCountries);
+	}
+
+	function isSourcePointDetectionDesktopEnabled(context) {
+		return context.opts.sourcePointDetectionUrl && (context.targeting.skin === 'oasis' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionCountries));
+	}
+
+	function isSourcePointDetectionMobile(context) {
+		return context.opts.sourcePointDetectionUrl && (context.targeting.skin === 'mercury' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionMobileCountries));
+	}
+
+	function updateRecoveryServicesAdContext(context, noExternals) {
+		var taboolaConfig = instantGlobals.wgAdDriverTaboolaConfig || {},
+			isRecoveryServiceEnabled = false;
+
+		// PageFair recovery
+		context.opts.pageFairRecovery = !noExternals && !isRecoveryServiceEnabled &&
+			context.opts.pageFairRecovery && geo.isProperGeo(instantGlobals.wgAdDriverPageFairRecoveryCountries);
+		isRecoveryServiceEnabled = context.opts.pageFairRecovery;
+
+		// PageFair detection
+		context.opts.pageFairDetection = !noExternals && isPageFairDetectionEnabled();
+
+		// SourcePoint recovery
+		context.opts.sourcePointRecovery = !noExternals && !isRecoveryServiceEnabled &&
+			context.opts.sourcePointRecovery && geo.isProperGeo(instantGlobals.wgAdDriverSourcePointRecoveryCountries);
+		isRecoveryServiceEnabled = isRecoveryServiceEnabled || context.opts.sourcePointRecovery;
+
+		// SourcePoint MMS
```

Comment:
what do you think about adding in code this tribal knowledge that we have in our minds and on confluence somewhere, that SP MMS can be enabled only when SP recovery is disabled? :) (documentation as close to code as possible)